### PR TITLE
Introduce trusted artifacts

### PR DESCRIPTION
**Pull request by zregvart**
Adds the use of trusted artifacts[1] to the Tasks and Pipelines. The `prefetch-dependencies` Task is no longer skipped based on the `hermetic` parameter -- it needs to run always to produce the trusted artifacts as the build* tasks depend on these, so if it were skipped the build* tasks would be skipped as well.

This also includes one small bugfix in the `buildah` task, it no longer tries to pull the `scratch` image.

Ref: https://issues.redhat.com/browse/EC-251

[1] https://github.com/redhat-appstudio/build-trusted-artifacts

### CODE REVIEW

File: task/buildah/0.1/buildah.yaml [hunk 1]
```diff
@@ -15,350 +16,354 @@ spec:
...
+      type: string
+    - default: ""
+      description: Delete image tag after specified time. Empty means to keep the image tag. Time values could be something like 1h, 2d, 3w for hours, days, and weeks, respectively.
+      name: IMAGE_EXPIRES_AFTER
+      type: string
+    - name: COMMIT_SHA
+      description: The image is built from this commit.
+      type: string
+      default: ""
+    - name: ARTIFACTS
```
Comment (lcarva) [inline task/buildah/0.1/buildah.yaml:61]: We probably want to create a new version of each Task that now task the ARTIFACTS parameter or emits an artifact result.

**Comment by stuartwdouglas** [comment]
It looks like you have accidentally formatted everything? There is a huge amount of non-related formatting changes in the PR.

File: task/buildah/0.1/buildah.yaml [hunk 2]
```diff
@@ -15,350 +16,354 @@ spec:
...
-        http_code=$(curl -s -L -w "%{http_code}" --output "$dockerfile_path" "$DOCKERFILE")
-        if [ $http_code != 200 ]; then
-          echo "No Dockerfile is fetched. Server responds $http_code"
+    - name: use-trusted-artifact
+      image: quay.io/redhat-appstudio/build-trusted-artifacts:latest@sha256:4581dc538a0bc1550466e8c7df3ee905e444eef3bc70a8ae765b568ab5a10588
+      args:
+        - use
+        - --store
+        - $(workspaces.source.path)
+        - $(params.ARTIFACTS[0])=$(workspaces.source.path)/source
```
Comment (stuartwdouglas) [inline task/buildah/0.1/buildah.yaml:103]: This kinda feels wrong to me. Wouldn't it make more sense to pass these in as explicit named parameters?

Rather than an array pram just use one parameter/result per artifact.

Comment (stuartwdouglas) [reply task/buildah/0.1/buildah.yaml:103]: Also this would remove the need to always run the prefetch step.

**Comment by stuartwdouglas** [comment]
Have you performance tested this? Last I heard we were IO bound on the PVCs, and this has the potential to add significantly more PVC IO, especially for builds with a large number of dependencies.

Is the plan to eventually move away from PVC storage altogether?

File: task/buildah/0.1/buildah.yaml (hunk 2 above)
Comment (zregvart) [reply task/buildah/0.1/buildah.yaml:103]: > This kinda feels wrong to me. Wouldn't it make more sense to pass these in as explicit named parameters?
> 
> Rather than an array pram just use one parameter/result per artifact.

The number of artifacts produced by a task is unbounded, e.g. the git-clone task [create one artifact](https://github.com/zregvart/build-definitions/blob/945441d9219ac4233a022da9ece34aa8ef2601d3/task/git-clone/0.1/git-clone.yaml#L269), whereas prefetch-dependencies [produces two](https://github.com/zregvart/build-definitions/blob/945441d9219ac4233a022da9ece34aa8ef2601d3/task/prefetch-dependencies/0.1/prefetch-dependencies.yaml#L77-L78).

> Also this would remove the need to always run the prefetch step.

I'm curious, can you elaborate a bit?

**Comment by zregvart** [comment]
> It looks like you have accidentally formatted everything? There is a huge amount of non-related formatting changes in the PR.

This was on purpose (see the [separate commit](https://github.com/redhat-appstudio/build-definitions/pull/712/commits/d612aaa8897845770083e1405250478db9134c95)), I would still like to avoid that and from what it seems the .dotfiles are not included in the archive.

> Have you performance tested this? Last I heard we were IO bound on the PVCs, and this has the potential to add significantly more PVC IO, especially for builds with a large number of dependencies.

No I have not, but I can do some preliminary performance tests. I'm pretty sure that the performance will not improve with this.

> Is the plan to eventually move away from PVC storage altogether?

Not something I can comment on, also not seeing how that would fit with Tekton's notion of [a workspace](https://tekton.dev/docs/pipelines/workspaces/), we would need to not use those to begin with.

File: task/buildah/0.1/buildah.yaml (hunk 1 above)
Comment (zregvart) [reply task/buildah/0.1/buildah.yaml:61]: > We probably want to create a new version of each Task that now task the ARTIFACTS parameter or emits an artifact result.

I'm not sold on this, if ARTIFACTS was made optional the backward compatibility would be maintained.

File: task/buildah/0.1/buildah.yaml (hunk 2 above)
Comment (stuartwdouglas) [reply task/buildah/0.1/buildah.yaml:103]: > > This kinda feels wrong to me. Wouldn't it make more sense to pass these in as explicit named parameters?
> > Rather than an array pram just use one parameter/result per artifact.
> 
> The number of artifacts produced by a task is unbounded, e.g. the git-clone task [create one artifact](https://github.com/zregvart/build-definitions/blob/945441d9219ac4233a022da9ece34aa8ef2601d3/task/git-clone/0.1/git-clone.yaml#L269), whereas prefetch-dependencies [produces two](https://github.com/zregvart/build-definitions/blob/945441d9219ac4233a022da9ece34aa8ef2601d3/task/prefetch-dependencies/0.1/prefetch-dependencies.yaml#L77-L78).

Each artifact is basically just a digest, so you would just name them, and then in the 'use' section specify the results you were reading from. E.g. git-clone would have a result called CHECKOUT_ARTIFACT, this would be pased in as a GIT_CHECKOUT param, and instead of `$(params.ARTIFACTS[0])=$(workspaces.source.path)/source` it would be `$(params.GIT_CHECKOUT=$(workspaces.source.path)/source`.

> 
> > Also this would remove the need to always run the prefetch step.
> 
> I'm curious, can you elaborate a bit?

I guess it would would still need to run, as I don't think Tekton allows subsequent tasks to run if a task is skipped, but you could abort early and write an empty file to the results, and have the artifact code ignore empty hashes.

Comment (zregvart) [reply task/buildah/0.1/buildah.yaml]: I think I can make `*_ARTIFACT` work, that will make the code easier to follow. I guess I was considering how we would tie the chain of artifacts easier in EC, having the same result/parameter name would help there.

File: task/buildah-remote/0.1/buildah-remote.yaml [hunk 3]
```diff
@@ -184,7 +201,9 @@ spec:
         BUILDAH_ARGS="--pull=never"
         UNSHARE_ARGS="--net"
         for image in $(grep -i '^\s*FROM' "$dockerfile_path" | sed 's/--platform=\S*//' | awk '{print $2}'); do
-          unshare -Ufp --keep-caps -r --map-users 1,1,65536 --map-groups 1,1,65536 -- buildah pull $image
+          if [ "${image}" != "scratch" ]; then
```
Comment (lcarva) [inline task/buildah-remote/0.1/buildah-remote.yaml:204]: This seems like an unrelated change? Same in `task/buildah/0.1/buildah.yaml`

Comment (stuartwdouglas) [reply task/buildah-remote/0.1/buildah-remote.yaml:204]: This is https://issues.redhat.com/browse/RHTAPBUGS-1066, maybe it should be moved to its own PR.

Comment (zregvart) [reply task/buildah-remote/0.1/buildah-remote.yaml:204]: Pulled into https://github.com/redhat-appstudio/build-definitions/pull/723

**Comment by lcarva** [comment]
@zregvart, what's preventing this from coming out of draft?

**Comment by zregvart** [comment]
> @zregvart, what's preventing this from coming out of draft?

I think I needed to performance test this, but I'm procrastinating...

**Comment by zregvart** [comment]
I'm looking for reviews here. And a decision if we can leave the performance testing for after. I'm not sure (Luiz seems to have some ideas though), how the potential impact on performance can be minimized. If any performance improvements are to be implemented they will not be implemented here but in the https://github.com/redhat-appstudio/build-trusted-artifacts repository.
With the [policy change](https://github.com/enterprise-contract/ec-policies/pull/868) in draft now I'm eager to get this work merged soon.

File: task/buildah/0.1/buildah.yaml [hunk 4]
```diff
@@ -105,6 +111,14 @@ spec:
...
       value: $(params.YUM_REPOS_D_TARGET)
   steps:
+  - name: use-trusted-artifact
+    image: quay.io/redhat-appstudio/build-trusted-artifacts:latest@sha256:7a6a70bb972dba913e3d087305eece1a3e0462fd5adae52cd32322b9997bf213
+    args:
+      - use
+      - --store
+      - $(workspaces.source.path)
+      - $(params.SOURCE_ARTIFACT)=$(workspaces.source.path)/source
+      - $(params.CACHI2_ARTIFACT)=$(workspaces.source.path)/cachi2
```
Comment (chmeliik) [inline task/buildah/0.1/buildah.yaml]: What happens if a user makes some modifications in the source directory in a custom task but doesn't use the trusted artifact mechanisms? I guess this step just throws away their modifications?

Would it be possible to warn the user or even fail the build when trusted-artifacts discards user modifications?

**Comment by chmeliik** [comment]
> Last I heard we were IO bound on the PVCs, and this has the potential to add significantly more PVC IO, especially for builds with a large number of dependencies.

A potential performance improvement could be to extract the trusted artifacts somewhere other than a PVC? `use --store <not-on-a-PVC>`

File: task/buildah/0.1/buildah.yaml (hunk 4 above)
Comment (lcarva) [reply task/buildah/0.1/buildah.yaml]: A couple of things may happen:

1. The user adds their custom task to the chain of tasks where it produces a new trusted artifact which is then wired up to the buildah task. In this case, the build will succeed but EC will catch this as a violation because an untrusted task modified the source.
2. The user adds a custom task but doesn't wire things up properly in the pipeline. Then, yes, the changes will just get thrown out.
3. The user adds a custom task that tries to modify the trusted artifact produced by a trusted task. This will cause a build failure when another task tries to use that archive.

**Comment by zregvart** [comment]
Moved out of draft because I'm anxious and ambitious at the same time.

**Comment by lcarva** [comment]
> > Last I heard we were IO bound on the PVCs, and this has the potential to add significantly more PVC IO, especially for builds with a large number of dependencies.
> 
> A potential performance improvement could be to extract the trusted artifacts somewhere other than a PVC? `use --store <not-on-a-PVC>`
> 
> Although that can be done as a follow-up if the performance does get worse

Yes! We hope to eventually start using OCI archives. But we (mostly me) didn't want to add that additional layer of complexity to this first iteration.

**Comment by chmeliik** [comment]
> Yes! We hope to eventually start using OCI archives. But we (mostly me) didn't want to add that additional layer of complexity to this first iteration.

Would it not be possible to do it without OCI archives? When creating the artifact, use `create --store <yes-on-a-PVC>`, when using it `use --store <not-on-a-PVC>`

For some filesystems, writing lots of small files is much worse than one large file, so the `create` step could be much cheaper than `use`.

_Flashbacks to the time when extracting a 1.0 GiB git repo archive to an aws-efs volume took 30 minutes_

**Comment by lcarva** [comment]
> > Yes! We hope to eventually start using OCI archives. But we (mostly me) didn't want to add that additional layer of complexity to this first iteration.
> 
> Would it not be possible to do it without OCI archives? When creating the artifact, use `create --store <yes-on-a-PVC>`, when using it `use --store <not-on-a-PVC>`
> 
> For some filesystems, writing lots of small files is much worse than one large file, so the `create` step could be much cheaper than `use`.
> 
> _Flashbacks to the time when extracting a 1.0 GiB git repo archive to an aws-efs volume took 30 minutes_

Oh! This reminds me of something. @zregvart, when a trusted artifact is being used, it should get extracted to an emptyDir that only the Task has access to. Otherwise, a Task running in parallel can change the contents after they've been verified.

**Comment by stuartwdouglas** [comment]
> > > Yes! We hope to eventually start using OCI archives. But we (mostly me) didn't want to add that additional layer of complexity to this first iteration.
> > 
> > 
> > Would it not be possible to do it without OCI archives? When creating the artifact, use `create --store <yes-on-a-PVC>`, when using it `use --store <not-on-a-PVC>`
> > For some filesystems, writing lots of small files is much worse than one large file, so the `create` step could be much cheaper than `use`.
> > _Flashbacks to the time when extracting a 1.0 GiB git repo archive to an aws-efs volume took 30 minutes_
> 
> Oh! This reminds me of something. @zregvart, when a trusted artifact is being used, it should get extracted to an emptyDir that only the Task has access to. Otherwise, a Task running in parallel can change the contents after they've been verified.

If we are worried about parallel tasks then shouldn't the 'store' task also do all its processing in an off-PVC location? At the moment it writes the archive to the store directory, the gets the digest and moves it/writes the results. If we are worried about parallel tasks then in theory a parallel task could modify the archive between creation and the original task generating the digest.

**Comment by zregvart** [comment]
> If we are worried about parallel tasks then shouldn't the 'store' task also do all its processing in an off-PVC location? At the moment it writes the archive to the store directory, the gets the digest and moves it/writes the results. If we are worried about parallel tasks then in theory a parallel task could modify the archive between creation and the original task generating the digest.

When restoring the archive from the store the hash of the archive is checked again, for this to be a concern a Task running in parallel would also need to be able to influence the `*_ARTIFACT` result of the Task creating the archive, I think that is possible only if the malicious Task also has permissions to modify the TaskRun.

**Comment by zregvart** [comment]
This will conflict with #771, so let's put this one on hold

**Comment by stuartwdouglas** [comment]
> > If we are worried about parallel tasks then shouldn't the 'store' task also do all its processing in an off-PVC location? At the moment it writes the archive to the store directory, the gets the digest and moves it/writes the results. If we are worried about parallel tasks then in theory a parallel task could modify the archive between creation and the original task generating the digest.
> 
> When restoring the archive from the store the hash of the archive is checked again, for this to be a concern a Task running in parallel would also need to be able to influence the `*_ARTIFACT` result of the Task creating the archive, I think that is possible only if the malicious Task also has permissions to modify the TaskRun.

I mean if you have two task running in parallel, with either `ReadWriteMany` or `ReadWriteOnce` but on the same node, then in this code here there is a potential for a race: https://github.com/redhat-appstudio/build-trusted-artifacts/blob/main/create.sh#L46

Basically at line 46 the archive has been created, but we have not yet calculated the digest. A malicious or erroneous task could in theory modify the archive at this point and we would not be able to detect it, as we have not yet calculated the sha.

**Comment by zregvart** [comment]
> I mean if you have two task running in parallel, with either `ReadWriteMany` or `ReadWriteOnce` but on the same node, then in this code here there is a potential for a race: [redhat-appstudio/build-trusted-artifacts@`main`/create.sh#L46](https://github.com/redhat-appstudio/build-trusted-artifacts/blob/main/create.sh?rgh-link-date=2024-01-31T10%3A08%3A41Z#L46)
> 
> Basically at line 46 the archive has been created, but we have not yet calculated the digest. A malicious or erroneous task could in theory modify the archive at this point and we would not be able to detect it, as we have not yet calculated the sha.

Only if the $TEMP directory is on the same PVC, right? That will not be the case.

**Comment by stuartwdouglas** [comment]
What do you mean by $TEMP? There is only a call to `mktemp -p ${store}`, which will create the temp file on the PVC pointed to by store. I think to avoid this you would need to call plain `mktemp`, create the archive in the /tmp dir, take the sha, then put them both onto the PVC.

**Comment by zregvart** [comment]
> What do you mean by $TEMP? There is only a call to `mktemp -p ${store}`, which will create the temp file on the PVC pointed to by store. I think to avoid this you would need to call plain `mktemp`, create the archive in the /tmp dir, take the sha, then put them both onto the PVC.

Right, thanks, I've created https://github.com/redhat-appstudio/build-trusted-artifacts/pull/14 to fix that. I guess the original intent was to make sure that the archive was created on a filesystem with enough free space. Let's see if that becomes an issue at any point.

**Comment by zregvart** [comment]
Submitted a followup in #791, please have a look. I'll close this one as the discussion here will not be applicable over there in most comments.
